feat:preload callbacks

nicomontanari opened the pull request:
I opened another [pull request](https://github.com/DylanVann/react-native-fast-image/pull/437) because the previous one was unexpectedly closed.

@DylanVann I have been using these changes in production for 1 year and everything works fine.

doomsower committed:
Add onProgress and onComplete callbacks for `preload` both on Android and iOS

doomsower committed:
Merge branch 'master' into preload_callbacks

# Conflicts:
#	android/src/main/java/com/dylanvann/fastimage/FastImagePreloaderModule.java
#	src/index.d.ts
#	src/index.js

Mickagd committed:
Merge remote-tracking branch 'banana/preload_callbacks' into feature/preload-callbacks

Mickagd committed:
fix flow

Elindorath committed:
fix: an empty sources array never called onComplete

Mickagd committed:
Merge pull request #1 from DylanVann/master

Update from original repository

Mickagd committed:
Merge branch 'master' into feature/preload-callbacks

Mickagd committed:
fix lint

Mickagd committed:
update swebimage

joan-saum committed:
Merge pull request #2 from DylanVann/master

Sync with the original repo

Commit:
Merge branch 'master' into feature/preload-callbacks

Commit:
Merge remote-tracking branch 'upstream/master' into feature/preload-callbacks-fetch-updates-2021-08-18

Commit:
fix: revert FastImagePreloaderModule class to FastImageViewModule + add callback to preload function

Commit:
fix lint

Commit:
fix NativeEventEmitter mock in test

Commit:
fix NativeEventEmitter mock in test

Commit:
turn preloadermanager.js to typescript file

Commit:
fix lint + add listener to FastImageViewModule.java

Commit:
bind context to listeners

Commit:
add SDWebImageDownloader.h import in FFFastImagePreloaderManager.m file

Mickagd committed:
Merge pull request #4 from Sparted/feature/preload-callbacks-fetch-updates-2021-08-18

Feature/preload callbacks fetch updates 2021 08 18

Flictuum committed:
Merge branch 'main' into feature/preload-callbacks

nicomontanari committed:
Import FastImageViewNativeModule into index file

nicomontanari committed:
Update FastImageViewNativeModule variable

nicomontanari committed:
Merge remote-tracking branch 'upstream/main'

mathbalduino commented:
Hey @DylanVann, it would be nice to have this one merged. What do you think?

lorenzpfei commented:
@DylanVann can you please merge this? It would be very helpful

akeva001 commented:
@nicomontanari would you be able to help me on adding this PR to my react native project? I have tried downloading the stable version of react-native-fast-image and applying the necessary changes to the files, but I am not able to get it working.

akeva001 commented:
> @nicomontanari would you be able to help me on adding this PR to my react native project? I have tried downloading the stable version of react-native-fast-image and applying the necessary changes to the files, but I am not able to get it working.

I ended up being able to install it properly I believe,
<img width="379" alt="Screenshot 2023-09-27 at 7 54 45 PM" src="https://github.com/DylanVann/react-native-fast-image/assets/57927148/3fc073a2-e03f-4b9e-b995-dcc405fa1334">

But this code block is not logging anything,
<img width="483" alt="Screenshot 2023-09-27 at 7 57 24 PM" src="https://github.com/DylanVann/react-native-fast-image/assets/57927148/daaaa5cc-d66d-4760-832b-518502273481">

sterlingwes commented:
This PR is missing the runtime change needed in `index.tsx`, it's still calling the old native module.

Instead replace with: 

```js
FastImage.preload = (sources, onProgress, onComplete) =>
    preloaderManager.preload(sources, onProgress, onComplete)
```

nicomontanari commented:
Hi @akeva001, i'm sorry for the delay in my response.

First i added this line in the "dependecies" section in the `package.json`: 
`"react-native-fast-image": "https://github.com/nicomontanari/react-native-fast-image"`.

Then this is how i implemented the preload:
```typescript
import FastImage from 'react-native-fast-image/src'

const preload = async (images: string[]): Promise<void> => {
    return new Promise<void>(async (resolve, reject) => {
        FastImage.preload(
            images.map((image) => ({
                uri: image
            })),
            (finished, total) => {
                if (__DEV__) {
                    console.log('Preload:', `f ${finished}/t ${total}`)
                }
            },
            (_, skipped) => {
                if (skipped > 0) {
                    reject(skipped)
                    if (__DEV__) {
                        console.log('Preload: failed with skipped ', skipped)
                    }
                }
                if (__DEV__) {
                    console.log('Preload: finish')
                }
                resolve()
            }
        )
    })
}
```

nicomontanari commented:
@sterlingwes i'm sorry but i didn't understand your comment. These changes already implement your piece of code.

sterlingwes commented:
@nicomontanari if you look in the PR diff here and scroll to the very bottom you'll see that you only have typescript changes in `src/index`. With no runtime change, the code snippet you shared above is still calling the existing native module method which doesn't do anything with the callback params.

You need to update the index file to pass those params into the method exposed by PreloaderManager which isn't currently used in this diff.

sterlingwes commented on android/src/main/java/com/dylanvann/fastimage/FastImagePreloaderListener.java:
```diff
@@ -0,0 +1,74 @@
+package com.dylanvann.fastimage;
+
+import android.support.annotation.Nullable;
```
```suggestion
import androidx.annotation.Nullable;
```

separate change also required here for android to build

nicomontanari committed:
Update android/src/main/java/com/dylanvann/fastimage/FastImagePreloaderListener.java

Co-authored-by: Wes Johnson <swj@wesquire.ca>

frg100 reviewed (commented):
It looks like `src/index.tsx:260` needs some changes to use the new interface

nicomontanari committed:
Update preload manager references

nicomontanari commented:
@frg100 @sterlingwes thanks for the tip, I missed these changes along the way 😕

raymondjacobson commented:
This would be really nice... what's the status?

frg100 reviewed (approved):
Thanks!
